crates/driver/src/infra/tokens.rs

```diff
@@ -0,0 +1,88 @@
+use {
+    crate::{domain::eth, infra::Ethereum},
+    anyhow::Result,
+    contracts::ERC20,
+    ethcontract::dyns::DynWeb3,
+    std::{
+        collections::HashMap,
+        sync::{Arc, RwLock},
+    },
+};
+
+#[derive(Clone, Debug, Default)]
+pub struct Metadata {
+    pub decimals: u8,
+    pub symbol: String,
+}
+
+/// Provides metadata of tokens.
+pub struct Fetcher {
+    web3: DynWeb3,
+    cache: Arc<RwLock<HashMap<eth::TokenAddress, Metadata>>>,
+}
+
+impl Fetcher {
+    pub fn new(eth: &Ethereum) -> Self {
+        Self {
+            web3: eth.web3().clone(),
+            cache: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Fetches metadata of the requested token from a node.
+    async fn fetch_token_infos(
+        &self,
+        addresses: &[eth::TokenAddress],
+    ) -> Vec<Result<(eth::TokenAddress, Metadata)>> {
+        let futures = addresses.iter().map(|address| async {
+            let erc20 = ERC20::at(&self.web3, address.0 .0);
+            // Use `try_join` because these calls get batched under the hood
+            // so if one of them fails the other will as well.
+            // Also this way we won't get incomplete data for a token.
+            let (decimals, symbol) = futures::future::try_join(
+                erc20.methods().decimals().call(),
+                erc20.methods().symbol().call(),
+            )
+            .await?;
+            Ok((*address, Metadata { decimals, symbol }))
+        });
+        futures::future::join_all(futures).await
+    }
+
+    /// Returns the `Metadata` for the given tokens. Note that the result will
+    /// not contain data for tokens that encountered errors while fetching
+    /// the data.
+    pub async fn get_token_infos(
```

Comment:
naming nit: I think that `token_info.get_token_infos()` is quite redundant in terms of naming. I would just call this method `fetch` or something, as then `token_info.fetch()` or `tokens.fetch` is, IMO, more descriptive.

Reply:
[a473160](https://github.com/cowprotocol/services/pull/1770/commits/a473160f0116036981fd94dbfe37532e46fc6cd8)